src/Logic.hs

```diff
@@ -526,33 +526,20 @@ handleMergeRequested projectConfig prId author state pr approvalType = do
     then pure $ Pr.setIntegrationStatus prId IncorrectBaseBranch state''
     else pure state''
 
-traversePullRequests
-  :: (PullRequestId -> PullRequest -> Action PullRequest)
-  -> ProjectState
-  -> Action ProjectState
-traversePullRequests f state = do
-  pullRequests' <- IntMap.traverseWithKey (\key -> f (PullRequestId key)) $ Pr.pullRequests state
-  pure state { Pr.pullRequests = pullRequests' }
-
 handleBuildStatusChanged :: Sha -> BuildStatus -> ProjectState -> Action ProjectState
-handleBuildStatusChanged buildSha newStatus =
-  traversePullRequests setBuildStatus
+handleBuildStatusChanged buildSha newStatus = pure . Pr.updatePullRequests setBuildStatus
   where
-  setBuildStatus prId pr =
-    case Pr.integrationStatus pr of
-      -- If there is an integration candidate, and its integration sha matches that of the build,
-      -- then update the build status for that pull request. Otherwise do nothing.
-      Integrated candidateSha _
-        | candidateSha == buildSha -> do
-            let pr' = pr { Pr.integrationStatus = Integrated buildSha newStatus }
-
-            -- Sometimes we want to leave an informative comment on the PR. This isn't for things
-            -- that require user feedback like build failures.
-            case newStatus of
-              BuildPending (Just url) -> pr' <$ leaveComment prId ("Waiting on CI job: " <> url)
-              _                       -> pure pr'
-
-      _ -> pure pr
+  setBuildStatus pr = case Pr.integrationStatus pr of
+    -- If there is an integration candidate, and its integration sha matches that of the build,
+    -- then update the build status for that pull request. Otherwise do nothing.
+    Integrated candidateSha _ | candidateSha == buildSha ->
+      pr { Pr.integrationStatus = Integrated buildSha newStatus
+         , Pr.needsFeedback = case newStatus of
+                              BuildPending (Just _) -> True
+                              BuildFailed _         -> True
```

Comment:
... (cont'd from previous comment), which would be here?

Reply:
Yes, exactly.  And we only need to post comments when we get the `BuildPending` status with a URL and when the build fails.